File: Problem29.java
package leetcode;

public class Problem29 {

    public static void main(String[] args) {
        Problem29 solution = new Problem29();
        System.out.println(solution.divide(3, 2));

    }

    public int divide(int dividend, int divisor) {
        int absX = Math.abs(dividend);
        int absY = Math.abs(divisor);
        boolean xIsMinus = dividend < 0 ? true : false;
        boolean yIsMinus = divisor < 0 ? true : false;
        if( (xIsMinus && yIsMinus) || (!xIsMinus && !yIsMinus)) {
            return divideNumbers(absX, absY, true);
        } else if(xIsMinus) {
            return divideNumbers(absX, divisor, false);
        } else {
            return divideNumbers(dividend, absY, false);
        }
    }
    public int divideNumbers(int dividend, int divisor, boolean isPositive) {
        int max = Integer.MAX_VALUE;
        int min = Integer.MIN_VALUE;
        if(dividend < divisor) return 0;
        int result = 0;
        if(dividend <= 0) {
            if(isPositive == true) return 1;
            else return -1;
        }
        if(isPositive) {
            if(result == max) return max;
            result = divideNumbers(dividend - divisor, divisor, isPositive) + 1 ;

        }
        else {

            if(result == min) return min;
            result = divideNumbers(dividend - divisor, divisor, isPositive) - 1 ;
        }
        return result;
    }
}

File: src/main/java/leetcode/problem29.c
int divide(int dividend, int divisor) {
   int result = 2;
    int shift = 0;
    int isPositive = 1;
    int negMax = -2147483648;
    int posMax = 2147483647;
    if(dividend < 0 && divisor < 0) {
    	isPositive = 0;
    	if(dividend > divisor) return 0;
    } else if(dividend >= 0 && divisor >= 0){
		isPositive = 0;
        if(dividend < divisor) return 0;
    } else {
        if(dividend > 0) dividend= ~dividend +1;
        if(divisor > 0) divisor= ~divisor +1;
        if(dividend > divisor) return 0;
    }
	
   if(dividend < 0 || divisor < 0) {
        int tmp = divisor<<shift;
        while(divisor<<shift >= dividend && divisor<<shift <= tmp) {
            tmp = divisor<<shift;
            shift++;
        }
   } else {
        int tmp = divisor<<shift;
        while(divisor<<shift <= dividend && divisor<<shift >= tmp) {
            tmp = divisor<<shift;
            shift++;
        }
   }
    int diff = 0 ;
    if(shift > 1) {
    	shift -=2;
    	result = result<< shift;
    	diff = dividend - (divisor<<shift+1);
	} else {
		result = 1;
		diff = dividend - divisor;
	}
    result =  result + divide(diff, divisor);
    if(isPositive == 0  && negMax == result) result = posMax;
    return isPositive == 1 ? result *(-1) : result;
}
